### PR TITLE
Daily extrapolation factor

### PR DESCRIPTION
**Pull request by jaihind213**
- apply extrapol/privacy at indexing phase
- introducing HIP - hidden in plain sight metrics (i.e. they are hidden but there, used to store original values of a metric)

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java [hunk 1]
```diff
@@ -93,6 +95,22 @@
...
+ * to apply masking at indexing time, in the job spec file set the 'jobProperties' to
+ *
+ * {
+ *   "type": "hadoop",
+ *           "partitionsSpec": {
+ *           "type": "hashed",
+ *           "targetPartitionSize": 5000000
+ *            },
+ *   "jobProperties": {
+ *      "mapred.child.java.opts": "-DDATASPARK_DO_MASKING_INDEX=true -Xmx200m -DDATASPARK_MASKED_METRICS=data4.count,data4.uniq,data4.sum",
```
Comment (vihag) [inline indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:107]: Can we put this somewhere in confluence ? Ill probably not be taking care of Druid stuff forever, everyone else needs to know

File: examples/bin/node.sh [hunk 2]
```diff
@@ -21,8 +21,23 @@ shift
...
+#masking zero counts can happen only if the flag 'DATASPARK_DO_MASKING' is set to true first, i.e. flag 'DATASPARK_DO_MASKING' takes precedence.
+#RUNTIME_MASKING_OPTS=" -DMASK_ZERO_COUNTS=false -DDATASPARK_DO_MASKING=true -DDATASPARK_EXTRAPOLATION_FACTOR=1.35 -DDATASPARK_PRIVACY_THRESHOLD=200 -DDATASPARK_MASKED_METRICS=data.uniq,data.special_count_metric,data.sum,data.count,data2.latencyMs,data2.uniq,data2.special_count_metric,data2.sum,data2.count "
+# or
+## (2) if REVEAL_HIP_METRICS_OPTS are set then that means masking has been done at indexing time, so dont use runtime masking and send the hipkey with post request to broker.
+#REVEAL_HIP_METRICS_OPTS=" -DREVEAL_HIP_METRIC=false -DHIP_METRIC_KEY=b275822c7da805d368bcd56fb614e95C# "
+
+if [ -n "$RUNTIME_MASKING_OPTS" ] && [ -n "$REVEAL_HIP_METRICS_OPTS" ]; then
+    echo "Both Runtime Masking and Revealing HIP metrics options have been set. Only one should be set."
+    echo "If metrics are masked at indexing time then use 'REVEAL_HIP_METRICS_OPTS' ."
```
Comment (vihag) [inline examples/bin/node.sh:37]: just curious here, what happens if midway we decide that we dont need runtime masking, is re-indexing the only way or can we switch the api to use the original metrics with runtime extrapol

File: indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java [hunk 3]
```diff
@@ -329,6 +354,9 @@ protected void setup(Context context)
       combiningAggs = new AggregatorFactory[aggregators.length];
       for (int i = 0; i < aggregators.length; ++i) {
         combiningAggs[i] = aggregators[i].getCombiningFactory();
+        TableDataSource tds = new TableDataSource(config.getDataSource());
```
Comment (vihag) [inline indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:357]: Is there a reason for re-initializing the datsource, combiner and aggregator ?

File: server/src/main/java/io/druid/server/HipMetricUtil.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ *  So in node.sh for the broker, set the following as part of JAVA_OPTS.
+ *  i.e.
+ *  JAVA_OPTS=" ${MASKING_OPTS} -DREVEAL_HIP_METRIC=true -DHIP_METRIC_KEY=b275822c7da805d368bcd56fb614e95C#
+ *  and restart broker.
+ *
+ *  when the broker receives a http post druid query, you should have http header 'BEHIP' and u must set it to the same reveal key to see HIP metrics.
+ */
+public class HipMetricUtil {
+
+  public static final String HIDDEN_METRIC_SUFFIX = "_hip_" ;
```
Comment (vihag) [inline server/src/main/java/io/druid/server/HipMetricUtil.java:45]: Is prefix/suffix a better way, or a configurable list of dimension names ?

File: server/src/main/java/io/druid/server/HipMetricUtil.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      canServerRevealHipMetrics();
+    } catch (RuntimeException e) {
+      System.out.println("** Server will NOT reveal metrics due to invalid reveal key.");
+      e.printStackTrace();
+    }
+  }
+
+
+  static boolean isHiddenMetric(String metricName) {
+    if (metricName == null || "".equalsIgnoreCase(metricName.trim())){
```
Comment (vihag) [inline server/src/main/java/io/druid/server/HipMetricUtil.java:68]: Use StringUtils, it was used in other code as well

**Review by vihag — APPROVED** [review]

**Comment by jaihind213** [comment]
you can ignore this pull request. we no longer need daily extpolation

**Pull request closed without merging**